No good deed goes unpunished, the new testing is bringing out some weird results:

<details>
<summary>Details</summary>

```
--- FAIL: TestKVNemesisSingleNode_ReproposalChaos (10.98s)
    test_log_scope.go:167: test logs captured to: /Users/tobias/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/3071ecfd607c2dcda710d9ffa8fdef40/logTestKVNemesisSingleNode_ReproposalChaos720228959
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:278: seed: 4043305725150643922
    kvnemesis_test.go:192: kvnemesis logging to /Users/tobias/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/3071ecfd607c2dcda710d9ffa8fdef40/kvnemesis3540977959
    kvnemesis.go:168: error applying x.Scan(ctx, tk(13779811691630448898), tk(17053492883382310021), 0) // failed to push meta={id=0d4951ff key=/Table/100/"d63f744e684afbd2" iso=ReadCommitted pri=0.04459242 epo=5 ts=1687351579.475751000,0 min=1687351579.468285000,0 seq=3} lock=true stat=PENDING rts=0,0 wto=false gul=0,0: {<nil> 0 {0x14000723090} index:0  1687351579.500642000,0}
    kvnemesis.go:180: could not scan actual latest values: injected error
        (1)
          | (opaque error wrapper)
          | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
          | reportable 0:
          |
          | github.com/cockroachdb/cockroach/pkg/kv/kvnemesis.init
          |     github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis.go:30
          | runtime.doInit
          |     GOROOT/src/runtime/proc.go:6349
          | runtime.doInit
          |     GOROOT/src/runtime/proc.go:6326
          | runtime.main
          |     GOROOT/src/runtime/proc.go:233
          | runtime.goexit
          |     GOROOT/src/runtime/asm_arm64.s:1172
        Wraps: (2) injected error
        Error types: (1) *errbase.opaqueWrapper (2) *errutil.leafError
    kvnemesis.go:188: failures(verbose): /Users/tobias/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/3071ecfd607c2dcda710d9ffa8fdef40/kvnemesis3540977959/failures
        repro steps: /Users/tobias/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/3071ecfd607c2dcda710d9ffa8fdef40/kvnemesis3540977959/repro.go
        rangefeed KVs: /Users/tobias/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/3071ecfd607c2dcda710d9ffa8fdef40/kvnemesis3540977959/kvs-rangefeed.txt
        scan KVs: 
    kvnemesis_test.go:305: 
                Error Trace:    github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:305
                                                        github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:242
                Error:          Should be zero, but was 1
                Test:           TestKVNemesisSingleNode_ReproposalChaos
                Messages:       kvnemesis detected failures
    panic.go:522: -- test log scope end --
```

</details>

I am not sure how `errInjected` even ends up in a _Scan_. Need to look at this.

--- a/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go
+++ b/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go
@@ -31,7 +31,8 @@ import (
 // The degree of the inFlightWrites btree.
 const txnPipelinerBtreeDegree = 32
 
-var pipelinedWritesEnabled = settings.RegisterBoolSetting(
+// PipelinedWritesEnabled is the kv.transaction.write_pipelining_enabled cluster setting.
+var PipelinedWritesEnabled = settings.RegisterBoolSetting(
 	settings.TenantWritable,
 	"kv.transaction.write_pipelining_enabled",
 	"if enabled, transactional writes are pipelined through Raft consensus",
@@ -432,7 +433,7 @@ func (tp *txnPipeliner) canUseAsyncConsensus(ctx context.Context, ba *kvpb.Batch
 		return false
 	}
 
-	if !pipelinedWritesEnabled.Get(&tp.st.SV) || tp.disabled {
+	if !PipelinedWritesEnabled.Get(&tp.st.SV) || tp.disabled {
 		return false
 	}
 

--- a/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner_test.go
+++ b/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner_test.go
@@ -858,7 +858,7 @@ func TestTxnPipelinerEnableDisableMixTxn(t *testing.T) {
 	tp, mockSender := makeMockTxnPipeliner(nil /* iter */)
 
 	// Start with pipelining disabled. Should NOT use async consensus.
-	pipelinedWritesEnabled.Override(ctx, &tp.st.SV, false)
+	PipelinedWritesEnabled.Override(ctx, &tp.st.SV, false)
 
 	txn := makeTxnProto()
 	keyA, keyC := roachpb.Key("a"), roachpb.Key("c")
@@ -885,7 +885,7 @@ func TestTxnPipelinerEnableDisableMixTxn(t *testing.T) {
 	require.Equal(t, 0, tp.ifWrites.len())
 
 	// Enable pipelining. Should use async consensus.
-	pipelinedWritesEnabled.Override(ctx, &tp.st.SV, true)
+	PipelinedWritesEnabled.Override(ctx, &tp.st.SV, true)
 
 	ba.Requests = nil
 	putArgs2 := kvpb.PutRequest{RequestHeader: kvpb.RequestHeader{Key: keyA}}
@@ -913,7 +913,7 @@ func TestTxnPipelinerEnableDisableMixTxn(t *testing.T) {
 
 	// Disable pipelining again. Should NOT use async consensus but should still
 	// make sure to chain on to any overlapping in-flight writes.
-	pipelinedWritesEnabled.Override(ctx, &tp.st.SV, false)
+	PipelinedWritesEnabled.Override(ctx, &tp.st.SV, false)
 
 	ba.Requests = nil
 	putArgs4 := kvpb.PutRequest{RequestHeader: kvpb.RequestHeader{Key: keyA}}

--- a/pkg/kv/kvserver/BUILD.bazel
+++ b/pkg/kv/kvserver/BUILD.bazel
@@ -465,6 +465,7 @@ go_test(
         "//pkg/util",
         "//pkg/util/admission",
         "//pkg/util/admission/admissionpb",
+        "//pkg/util/allstacks",
         "//pkg/util/caller",
         "//pkg/util/circuit",
         "//pkg/util/ctxgroup",

--- a/pkg/kv/kvserver/replica_application_result.go
+++ b/pkg/kv/kvserver/replica_application_result.go
@@ -164,7 +164,16 @@ func (r *Replica) prepareLocalResult(ctx context.Context, cmd *replicatedCmd) {
 			//
 			// These many possible worlds are a major source of complexity, a
 			// reduction of which is postponed.
-			pErr = r.tryReproposeWithNewLeaseIndex(ctx, cmd)
+			pErr = nil
+			if fn := r.store.TestingKnobs().InjectReproposalError; fn != nil {
+				if err := fn(cmd.proposal); err != nil {
+					pErr = kvpb.NewError(err)
+				}
+			}
+			if pErr == nil { // since we might have injected an error
+				pErr = r.tryReproposeWithNewLeaseIndex(ctx, cmd)
+			}
+
 			if pErr != nil {
 				// An error from tryReproposeWithNewLeaseIndex implies that the current
 				// entry is not superseded (i.e. we don't have a reproposal at a higher

--- a/pkg/kv/kvserver/replica_raft_test.go
+++ b/pkg/kv/kvserver/replica_raft_test.go
@@ -11,21 +11,38 @@
 package kvserver
 
 import (
+	"context"
+	"fmt"
 	"path/filepath"
+	"strings"
+	"sync/atomic"
 	"testing"
 	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/logstore"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/datapathutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/echotest"
+	"github.com/cockroachdb/cockroach/pkg/util/allstacks"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
+	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/pebble"
 	"github.com/cockroachdb/redact"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 	"go.etcd.io/raft/v3/tracker"
 )
 
@@ -123,4 +140,244 @@ func Test_handleRaftReadyStats_SafeFormat(t *testing.T) {
 
 	echotest.Require(t, string(redact.Sprint(stats)),
 		filepath.Join(datapathutils.TestDataPath(t, "handle_raft_ready_stats.txt")))
+}
+
+// TestProposalsWithInjectedLeaseIndexAndReproposalError runs a number of
+// increment operations whose proposals are randomly subjected to invalid
+// LeaseAppliedIndex assignments, and for which reproposals randomly fail with
+// an injected error. The test verifies that increments that "went through" are
+// actually present (and not duplicated) and that operations that failed with an
+// unambiguous error did not apply.
+func TestProposalsWithInjectedLeaseIndexAndReproposalError(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	testutils.RunTrueAndFalse(t, "pipelined", testProposalsWithInjectedLeaseIndexAndReproposalError)
+}
+
+func testProposalsWithInjectedLeaseIndexAndReproposalError(t *testing.T, pipelined bool) {
+	stopper := stop.NewStopper()
+	ctx := context.Background()
+
+	tc := testContext{}
+
+	isOurCommand := func(ba *kvpb.BatchRequest) (_ string, ok bool) {
+		if ba == nil {
+			return "", false // not local proposal
+		}
+		inc := ba.Requests[0].GetIncrement()
+		if inc == nil {
+			return "", false
+		}
+		key := string(inc.Key)
+		if !strings.HasSuffix(key, "-testing") {
+			return "", false
+		}
+		return key, true
+	}
+
+	rnd, seed := randutil.NewPseudoRand()
+	t.Logf("seed: %d", seed)
+
+	shouldInject := func(baseProb float64, attempt int) bool { // start at attempt one
+		// Example: baseProb = 0.8
+		// On attempt 1, 80% chance of catching retry.
+		// On attempt 2, 40%.
+		// On attempt 3, 27%.
+		// And so on.
+		thresh := baseProb / float64(attempt)
+		return rnd.Float64() < thresh
+	}
+
+	cfg := TestStoreConfig(hlc.NewClockForTesting(nil))
+
+	var injectedReproposalErrors atomic.Int32
+	{
+		var mu syncutil.Mutex
+		seen := map[string]int{} // access from proposal buffer under raftMu
+		cfg.TestingKnobs.InjectReproposalError = func(p *ProposalData) error {
+			key, ok := isOurCommand(p.Request)
+			if !ok {
+				return nil
+			}
+			mu.Lock()
+			defer mu.Unlock()
+			seen[key]++
+			if !shouldInject(0.2, seen[key]) {
+				return nil
+			}
+			t.Logf("inserting reproposal error for %s (seen %d times)", key, seen[key])
+			return errors.Errorf("injected error")
+		}
+	}
+
+	var insertedIllegalLeaseIndex atomic.Int32
+	{
+		var mu syncutil.Mutex
+		seen := map[string]int{}
+		cfg.TestingKnobs.LeaseIndexFilter = func(p *ProposalData) kvpb.LeaseAppliedIndex {
+			key, ok := isOurCommand(p.Request)
+			if !ok {
+				return 0
+			}
+			mu.Lock()
+			seen[key]++
+			defer mu.Unlock()
+			if !shouldInject(0.8, seen[key]) { // very frequent reproposals
+				return 0
+			}
+			insertedIllegalLeaseIndex.Add(1)
+			// LAI 1 is always going to fail because the LAI is initialized when the lease
+			// comes into existence. (It's important that we pick one here that reliably
+			// fails because otherwise we may accidentally regress the closed timestamp[^1].
+			//
+			// [^1]: https://github.com/cockroachdb/cockroach/issues/70894#issuecomment-1433244880
+			return 1
+		}
+	}
+
+	tc.StartWithStoreConfig(ctx, t, stopper, cfg)
+
+	defer checkNoLeakedTraceSpans(t, tc.store)
+	defer stopper.Stop(ctx)
+
+	tc.store.cfg.Tracer().SetActiveSpansRegistryEnabled(true)
+	tc.store.cfg.Tracer().PanicOnUseAfterFinish()
+
+	kvcoord.PipelinedWritesEnabled.Override(ctx, &tc.store.ClusterSettings().SV, pipelined)
+
+	k := func(idx int) roachpb.Key {
+		var key roachpb.Key
+		key = append(key, tc.repl.Desc().StartKey...)
+		key = append(key, fmt.Sprintf("%05d", idx)...)
+		key = append(key, "-testing"...)
+		return key
+	}
+	var observedAsyncWriteFailures atomic.Int32
+	var observedReproposalErrors atomic.Int32
+	const iters = 300
+	expectations := map[string]int{}
+	for i := 0; i < iters; i++ {
+		key := k(i)
+
+		if err := tc.store.DB().Txn(ctx, func(ctx context.Context, txn *kv.Txn) (err error) {
+			defer func() {
+				if err != nil {
+					t.Logf("%d: error: %s", i, err)
+					var txnErr *kvpb.TransactionRetryWithProtoRefreshError
+					// NB: the string matching here is sad but the error has been
+					// flattened away at this point.
+					ignored := true
+					switch {
+					case errors.As(err, &txnErr) && strings.Contains(txnErr.String(), `missing intent on`):
+						observedAsyncWriteFailures.Add(1)
+					case testutils.IsError(err, `injected error`):
+						observedReproposalErrors.Add(1)
+					case errors.HasType(err, (*kvpb.AmbiguousResultError)(nil)):
+						t.Logf("%d: ignoring: %s", i, err)
+					default:
+						ignored = false
+					}
+					if ignored {
+						t.Logf("%d: ignoring: %s", i, err)
+					} else {
+						t.Errorf("%d: %s", i, err)
+					}
+				}
+			}()
+			// The handling of trace spans during reproposals is tricky,
+			// so make sure everyone is tracing to have the best chance
+			// at causing problems. Note that the tracer was configured
+			// to always trace and also to debug use-after-finish.
+			ctx, sp := tc.store.cfg.Tracer().StartSpanCtx(ctx, fmt.Sprintf("op %d", i))
+			defer sp.Finish()
+
+			// We write and delete the key first. If there were a bug in the
+			// reproposal pipeline, it could happen that the increment below and this
+			// rewrite cycle got reordered, and we would notice that the key didn't
+			// have the right value in the end. (The Put shouldn't make a difference
+			// but it removes the question that may arise in the mind of a reader of
+			// whether the Delete actually causes a write). Another benefit of this
+			// sequence is that if we leaked any latches, the test would stall, since
+			// we are mutating the same key repeatedly.
+			if err := txn.Put(ctx, key, "hello"); err != nil {
+				return err
+			}
+			if _, err := txn.Del(ctx, key); err != nil {
+				return err
+			}
+
+			resp, err := txn.Inc(ctx, key, 1)
+			if err != nil {
+				return err
+			}
+			n, err := resp.Value.GetInt()
+			if err != nil {
+				return err
+			}
+			if n != 1 {
+				return errors.Errorf("%d: got %d", i, n)
+			}
+			return txn.Commit(ctx)
+		}); err == nil {
+			expectations[string(key)] = 1
+		} else if !errors.HasType(err, (*kvpb.AmbiguousResultError)(nil)) {
+			expectations[string(key)] = 0
+		}
+	}
+
+	for keyString, exp := range expectations {
+		keyVal, err := tc.store.DB().Get(ctx, roachpb.Key(keyString))
+		require.NoError(t, err)
+		if exp == 0 {
+			require.Nil(t, keyVal.Value)
+			continue
+		}
+		n, err := keyVal.Value.GetInt()
+		require.NoError(t, err)
+		require.EqualValues(t, exp, n)
+	}
+	t.Logf("observed %d async write restarts, observed %d/%d injected aborts, %d injected illegal lease applied indexes",
+		observedAsyncWriteFailures.Load(), observedReproposalErrors.Load(), injectedReproposalErrors.Load(), insertedIllegalLeaseIndex.Load())
+	if pipelined {
+		// If we did pipelined writes, if we needed to repropose and injected an
+		// error, this should surface as an async write failure instead.
+		require.Zero(t, observedReproposalErrors.Load())
+	}
+	if !pipelined {
+		// If we're not pipelined, we shouldn't be able to get an async write
+		// failure. This isn't testing anything about reproposals per se, rather
+		// it's validation that we're truly not doing pipelined writes.
+		require.Zero(t, observedAsyncWriteFailures.Load())
+	}
+}
+
+func checkNoLeakedTraceSpans(t *testing.T, store *Store) {
+	testutils.SucceedsSoon(t, func() error {
+		var sps []tracing.RegistrySpan
+		_ = store.cfg.Tracer().VisitSpans(func(span tracing.RegistrySpan) error {
+			sps = append(sps, span)
+			return nil
+		})
+		if len(sps) == 0 {
+			return nil
+		}
+		var buf strings.Builder
+		fmt.Fprintf(&buf, "unexpectedly found %d active spans:\n", len(sps))
+		var ids []uint64
+		for _, sp := range sps {
+			trace := sp.GetFullRecording(tracingpb.RecordingVerbose)
+			for _, rs := range trace.Flatten() {
+				// NB: it would be a sight easier to just include these in the output of
+				// the string formatted recording, but making a change there presumably requires
+				// lots of changes across various testdata in the codebase and the author is
+				// trying to be nimble.
+				ids = append(ids, rs.GoroutineID)
+			}
+			fmt.Fprintln(&buf, trace)
+			fmt.Fprintln(&buf)
+		}
+		sl := allstacks.Get()
+		return errors.Newf("%s\n\ngoroutines of interest: %v\nstacks:\n\n%s", buf.String(), ids, sl)
+	})
 }

--- a/pkg/kv/kvserver/testing_knobs.go
+++ b/pkg/kv/kvserver/testing_knobs.go
@@ -493,6 +493,10 @@ type StoreTestingKnobs struct {
 	// required.
 	BaseQueueInterceptor func(ctx context.Context, bq *baseQueue)
 
+	// InjectReproposalError injects an error in tryReproposeWithNewLeaseIndex.
+	// If nil is returned, reproposal will be attempted.
+	InjectReproposalError func(p *ProposalData) error
+
 	// FlowControlTestingKnobs provide fine-grained control over the various
 	// kvflowcontrol components for testing.
 	FlowControlTestingKnobs *kvflowcontrol.TestingKnobs

